new-dti-website/components/courses/Timeline.tsx

```diff
@@ -0,0 +1,222 @@
+   *
+   * @returns A number representing the percentage of progress (from 0 to 100).
+   *
+   * @example
+   * ```ts
+   * const progressPercentage = getProgressPercentage();
+   * ```
+   */ const getProgressPercentage = () => {
+    const sortedEvents = events.sort((a, b) => {
+      const aDate = parseEventDate(a.date, a.time);
+      const bDate = parseEventDate(b.date, b.time);
+      return aDate.getTime() - bDate.getTime();
+    });
+
+    const firstEventDate = parseEventDate(sortedEvents[0].date, sortedEvents[0].time);
+    const lastEventDate = parseEventDate(
+      sortedEvents[sortedEvents.length - 1].date,
+      sortedEvents[sortedEvents.length - 1].time
+    );
+
+    const totalTimeSpan = lastEventDate.getTime() - firstEventDate.getTime();
+    const timeElapsed = Math.max(0, currentDate.getTime() - firstEventDate.getTime());
+    const progress = (timeElapsed / totalTimeSpan) * 100;
+
+    return Math.min(100, Math.max(0, progress));
+  };
+
+  const progressPercentage = getProgressPercentage();
+
+  // * Use Layout Effect to Handle resizing of the Timeline
+  useLayoutEffect(() => {
+    /**
+     * Calculates the length of the line connecting the first and last events.
+     *
+     * @remarks
+     * This function determines the position of the first and last event elements
+     * on the page and calculates the distance between them, either vertically
+     * (on mobile) or horizontally (on desktop). It updates the line length state accordingly.
+     *
+     * @returns A number representing the length of the line in pixels.
+     *
+     * @example
+     * ```ts
+     * calculateLineLength();
+     * ```
+     */
+    const calculateLineLength = () => {
+      if (events.length === 1 || events.length === 0) {
+        setLineLength(0);
+        return;
+      }
+      if (firstEventRef.current && lastEventRef.current) {
+        const firstPos = firstEventRef.current.getBoundingClientRect();
+        const lastPos = lastEventRef.current.getBoundingClientRect();
+
+        const firstCenter = {
+          x: firstPos.left + firstPos.width / 2,
+          y: firstPos.top + firstPos.height / 2
+        };
+
+        const lastCenter = {
+          x: lastPos.left + lastPos.width / 2,
+          y: lastPos.top + lastPos.height / 2
+        };
+
+        if (isMobile) {
+          const verticalDistance = Math.abs(lastCenter.y - firstCenter.y);
+          setLineLength(verticalDistance);
+        } else {
+          const horizontalDistance = Math.abs(lastCenter.x - firstCenter.x);
+          setLineLength(horizontalDistance);
+        }
+      }
+    };
+    const handleResize = () => {
+      const mobile = window.innerWidth < 640;
+      setIsMobile(mobile);
+      calculateLineLength();
+    };
+
+    handleResize();
+    window.addEventListener('resize', handleResize);
+
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  }, [isMobile, lineLength]);
```

Comment:
You can use `useScreenSize`, in 'src/hooks'. And maybe add 640 to `consts.ts` as a breakpoint.

Reply:
I didn't see this thanks Chris!